optuna/multi_objective/samplers/_nsga2.py

```diff
@@ -0,0 +1,256 @@
+
+        if not (0.0 <= mutation_prob <= 1.0):
+            raise ValueError("`mutation_prob` must be a float value within the range [0.0, 1.0].")
+
+        if not (0.0 <= crossover_prob <= 1.0):
+            raise ValueError("`crossover_prob` must be a float value within the range [0.0, 1.0].")
+
+        self._population_size = population_size
+        self._crossover_prob = crossover_prob
+        self._mutation_prob = mutation_prob
+        self._random_sampler = multi_objective.samplers.RandomMultiObjectiveSampler(seed=seed)
+        self._rng = np.random.RandomState(seed)
+
+    def infer_relative_search_space(
+        self,
+        study: "multi_objective.study.MultiObjectiveStudy",
+        trial: "multi_objective.trial.FrozenMultiObjectiveTrial",
+    ) -> Dict[str, BaseDistribution]:
+        return {}
+
+    def sample_relative(
+        self,
+        study: "multi_objective.study.MultiObjectiveStudy",
+        trial: "multi_objective.trial.FrozenMultiObjectiveTrial",
+        search_space: Dict[str, BaseDistribution],
+    ) -> Dict[str, Any]:
+        parent_generation, population = self._collect_parent_population(study)
+        trial_id = trial._trial_id
+
+        if len(population) == 0:
+            generation = 0
+            study._storage.set_trial_system_attr(trial_id, _GENERATION_KEY, generation)
+        else:
+            p0 = self._select_parent(study, population)
+            p1 = self._select_parent(study, population)
+
+            generation = parent_generation + 1
+            study._storage.set_trial_system_attr(trial_id, _GENERATION_KEY, generation)
+            study._storage.set_trial_system_attr(
+                trial_id, _PARENTS_KEY, [p0._trial_id, p1._trial_id]
+            )
+
+        return {}
+
+    def sample_independent(
+        self,
+        study: "multi_objective.study.MultiObjectiveStudy",
+        trial: "multi_objective.trial.FrozenMultiObjectiveTrial",
+        param_name: str,
+        param_distribution: BaseDistribution,
+    ) -> Any:
+        if _PARENTS_KEY not in trial.system_attrs:
+            return self._random_sampler.sample_independent(
+                study, trial, param_name, param_distribution
+            )
+
+        p0, p1 = trial.system_attrs[_PARENTS_KEY]
+
+        param = study._storage.get_trial(p0).params.get(param_name, None)
+        if param is None or self._rng.rand() < self._crossover_prob:
+            param = study._storage.get_trial(p1).params.get(param_name, None)
+
+        if param is None or self._rng.rand() < self._mutation_prob:
+            return self._random_sampler.sample_independent(
+                study, trial, param_name, param_distribution
+            )
+
+        return param
+
+    def _collect_parent_population(
+        self, study: "multi_objective.study.MultiObjectiveStudy"
+    ) -> Tuple[int, List["multi_objective.trial.FrozenMultiObjectiveTrial"]]:
+        # TODO(ohta): Optimize this method.
```

Comment:
> an easy thing you could do is to replace the trials list with a dictionary {"generation": "trials"} not having to go through all trials in each outer loop.

Sounds good. I adopted the approach at the commit https://github.com/optuna/optuna/pull/1163/commits/c0c3eff0a4dfb697cb1c76f55bd33a3d3946ff86.

> Just wondering but do you have any concrete ideas?

I thought to introduce a caching mechanism like shown below:
```diff
diff --git a/optuna/multi_objective/samplers/_nsga2.py b/optuna/multi_objective/samplers/_nsga2.py
index 17176939..ffbc76ff 100644
--- a/optuna/multi_objective/samplers/_nsga2.py
+++ b/optuna/multi_objective/samplers/_nsga2.py
@@ -1,4 +1,5 @@
 from collections import defaultdict
+import hashlib
 import itertools
 from typing import Any
 from typing import DefaultDict
@@ -19,6 +20,7 @@ from optuna.multi_objective.samplers import BaseMultiObjectiveSampler
 # Define key names of `Trial.system_attrs`.
 _GENERATION_KEY = "multi_objective:nsga2:generation"
 _PARENTS_KEY = "multi_objective:nsga2:parents"
+_POPULATION_KEY = "multi_objective:nsga2:population:{}"
 
 
 @experimental("1.5.0")
@@ -134,19 +136,27 @@ class NSGAIIMultiObjectiveSampler(BaseMultiObjectiveSampler):
     def _collect_parent_population(
         self, study: "multi_objective.study.MultiObjectiveStudy"
     ) -> Tuple[int, List["multi_objective.trial.FrozenMultiObjectiveTrial"]]:
-        # TODO(ohta): Optimize this method.
-
+        trials = study.get_trials(deepcopy=False)
+        generation_to_runnings = defaultdict(list)
         generation_to_population = defaultdict(list)
-        for trial in study.get_trials(deepcopy=False):
+        for trial in trials:
+            if _GENERATION_KEY not in trial.system_attrs:
+                continue
+
+            generation = trial.system_attrs[_GENERATION_KEY]
+
             if trial.state != optuna.trial.TrialState.COMPLETE:
+                if trial.state == optuna.trial.TrialState.RUNNING:
+                    generation_to_runnings[generation].append(trial.number)
                 continue
 
-            generation = trial.system_attrs.get(_GENERATION_KEY, 0)
             generation_to_population[generation].append(trial)
 
+        hasher = hashlib.sha256()
         parent_population = []  # type: List[multi_objective.trial.FrozenMultiObjectiveTrial]
         parent_generation = -1
-        for generation in itertools.count():
+        while True:
+            generation = parent_generation + 1
             population = generation_to_population[generation]
 
             # Under multi-worker settings, the population size might become larger than
@@ -154,22 +164,47 @@ class NSGAIIMultiObjectiveSampler(BaseMultiObjectiveSampler):
             if len(population) < self._population_size:
                 break
 
+            for trial in generation_to_runnings[generation]:
+                hasher.update(bytes(str(trial.number), "utf-8"))
+
+            population_cache_key = _POPULATION_KEY.format(hasher.hexdigest())
+            parent_generation, parent_population_numbers = study.system_attrs.get(
+                population_cache_key, (-1, [])
+            )
+            if parent_generation >= generation:
+                parent_population = [trials[n] for n in parent_population_numbers]
+                continue
+
             population.extend(parent_population)
-            parent_population = []
+            parent_population = self._select_elite_population(study, population)
             parent_generation = generation
 
-            population_per_rank = _fast_non_dominated_sort(population, study.directions)
-            for population in population_per_rank:
-                if len(parent_population) + len(population) < self._population_size:
-                    parent_population.extend(population)
-                else:
-                    n = self._population_size - len(parent_population)
-                    _crowding_distance_sort(population)
-                    parent_population.extend(population[:n])
-                    break
+            if len(generation_to_runnings[parent_generation]) == 0:
+                study.set_system_attr(
+                    population_cache_key,
+                    (parent_generation, [t.number for t in parent_population]),
+                )
 
         return parent_generation, parent_population
 
+    def _select_elite_population(
+        self,
+        study: "multi_objective.study.MultiObjectiveStudy",
+        population: List["multi_objective.trial.FrozenMultiObjectiveTrial"],
+    ) -> List["multi_objective.trial.FrozenMultiObjectiveTrial"]:
+        elite_population = []
+        population_per_rank = _fast_non_dominated_sort(population, study.directions)
+        for population in population_per_rank:
+            if len(elite_population) + len(population) < self._population_size:
+                elite_population.extend(population)
+            else:
+                n = self._population_size - len(elite_population)
+                _crowding_distance_sort(population)
+                elite_population.extend(population[:n])
+                break
+
+        return elite_population
+
     def _select_parent(
         self,
         study: "multi_objective.study.MultiObjectiveStudy",
```

I'd like the comment to keep as it is because I'll create a PR to introduce the cache just after this PR is merged.
(To summarize the above logic in a sentence is a bit hard to me...)